File: SeLeAs/LeetCode/LeetCode 75/totalCost.py
from typing import List
import heapq

class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        n = len(costs)
        total_cost = 0

        leftHeap = []
        rightHeap = []

        left = 0
        right = n - 1

        # Initialize both heaps
        for _ in range(candidates):
            if left <= right:
                heapq.heappush(leftHeap, (costs[left], left))
                left += 1
        for _ in range(candidates):
            if left <= right:
                heapq.heappush(rightHeap, (costs[right], right))
                right -= 1

        while k > 0:
            # Choose from the smaller-cost heap
            if leftHeap and (not rightHeap or leftHeap[0][0] <= rightHeap[0][0]):
                cost, idx = heapq.heappop(leftHeap)
                if left <= right:
                    heapq.heappush(leftHeap, (costs[left], left))
                    left += 1
            else:
                cost, idx = heapq.heappop(rightHeap)
                if left <= right:
                    heapq.heappush(rightHeap, (costs[right], right))
                    right -= 1

            total_cost += cost
            k -= 1

        return total_cost

File: Self_Learning_Assignments/LeetCode/LeetCode_75/totalCost.java
import java.util.*;
public class totalCost {


public class Solution {
    public long totalCost(int[] costs, int k, int candidates) {
        int n = costs.length;
        long totalCost = 0;

        PriorityQueue<int[]> leftHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        PriorityQueue<int[]> rightHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));

        int left = 0, right = n - 1;

        // Fill the left and right heaps with up to 'candidates' elements
        for (int i = 0; i < candidates && left <= right; i++) {
            leftHeap.offer(new int[]{costs[left], left});
            left++;
        }

        for (int i = 0; i < candidates && left <= right; i++) {
            rightHeap.offer(new int[]{costs[right], right});
            right--;
        }

        while (k > 0) {
            if (!leftHeap.isEmpty() && (rightHeap.isEmpty() || leftHeap.peek()[0] <= rightHeap.peek()[0])) {
                int[] worker = leftHeap.poll();
                totalCost += worker[0];

                if (left <= right) {
                    leftHeap.offer(new int[]{costs[left], left});
                    left++;
                }
            } else {
                int[] worker = rightHeap.poll();
                totalCost += worker[0];

                if (left <= right) {
                    rightHeap.offer(new int[]{costs[right], right});
                    right--;
                }
            }
            k--;
        }

        return totalCost;
    }
}

}
